main/say.c

```diff
@@ -353,26 +354,29 @@ static int say_digit_str_full(struct ast_channel *chan, const char *str, const c
 static struct ast_str* ast_get_money_en_dollars_str(const char *str, const char *lang)
 {
 	const char *fnr;
-
-	double dollars = 0;
-	int amt, cents;
+	int amt, dollars = 0, cents = 0;
 	struct ast_str *fnrecurse = NULL;
+	struct ast_str *filenames;
 
-	struct ast_str *filenames = ast_str_create(20);
+	if (ast_strlen_zero(str)) {
+		return NULL;
+	}
+
+	filenames = ast_str_create(20);
 	if (!filenames) {
 		return NULL;
 	}
 	ast_str_reset(filenames);
 
-	if (sscanf(str, "%30lf", &dollars) != 1) {
+	/* Don't use %f because floating point rounding
+	 * could distort the cents units. Just parse as string. */
+	if (sscanf(str, "%d.%02d", &dollars, &cents) < 1) {
```

Comment:
Should be allowed: `.25`
Should fail: `1.-25` and `1.    25`

Reply:
Thanks, covered these in some additional test cases